src/perl/lib/wtsi_clarity/epp/sm/report_maker.pm

```diff
@@ -418,28 +421,42 @@ sub _build__all_udf_values {
   return $data;
 }
 
-sub _get_artifact_uris_from_udf {
-  my ($self, $step, $udf_name) = @_;
+sub _search_artifacts {
+  my ($self, $step, $udf_name, $start_index) = @_;
 
-  my $res_arts_doc = $self->request->query_resources(
+  return $self->request->query_resources(
         q{artifacts},
         {
-          udf       => qq{udf.$udf_name.min=0},
-          type      => qq{Analyte},
-          step      => $step,
-          sample_id => $self->_sample_ids(),
+          udf         => qq{udf.$udf_name.min=0},
+          type        => qq{Analyte},
+          step        => $step,
+          sample_id   => $self->_sample_ids(),
+          start_index => $start_index // 0,
         }
   );
-
-  return $self->grab_values($res_arts_doc, $ARTEFACTS_ARTEFACT_URIS_PATH);
 };
 
 sub _get_udf_values {
   my ($self, $step, $udf_name) = @_;
 
-  my $artifacts_uris = $self->_get_artifact_uris_from_udf($step, $udf_name);
-  my $artifacts = $self->request->batch_retrieve('artifacts', $artifacts_uris);
-  my @nodes;
+  my $artifacts_doc;
+  my @artifacts_uris = ();
+  my $start_index = 0;
+
+  do {
+    $artifacts_doc  = $self->_search_artifacts($step, $udf_name, $start_index);
+    push @artifacts_uris, @{$self->grab_values($artifacts_doc, $ARTEFACTS_ARTEFACT_URIS_PATH)};
+
+    $start_index += $PAGE_SIZE;
+
+  } ## no critic(ControlStructures::ProhibitPostfixControls)
+    while ($artifacts_doc->findvalue($NEXT_PAGE_URI) ne q{});
```

Comment:
Okay, I will change it. 

But how do I make it shorter? I thought it'd be longer as I have to do the search first, then do more searches inside the loop if there are more pages to fetch.

On 27 Apr 2015, at 11:38, Károly Erdős <notifications@github.com> wrote:

> In src/perl/lib/wtsi_clarity/epp/sm/report_maker.pm:
> 
> >  };
> >  
> >  sub _get_udf_values {
> >    my ($self, $step, $udf_name) = @_;
> >  
> > -  my $artifacts_uris = $self->_get_artifact_uris_from_udf($step, $udf_name);
> > -  my $artifacts = $self->request->batch_retrieve('artifacts', $artifacts_uris);
> > -  my @nodes;
> > +  my $artifacts_doc;
> > +  my @artifacts_uris = ();
> > +  my $start_index = 0;
> > +
> > +  do {
> > +    $artifacts_doc  = $self->_search_artifacts($step, $udf_name, $start_index);
> > +    push @artifacts_uris, @{$self->grab_values($artifacts_doc, $ARTEFACTS_ARTEFACT_URIS_PATH)};
> > +
> > +    $start_index += $PAGE_SIZE;
> > +
> > +  } ## no critic(ControlStructures::ProhibitPostfixControls)
> > +    while ($artifacts_doc->findvalue($NEXT_PAGE_URI) ne q{});
> Yes, I understand it, but I personally do not like turning on/off critics, if it is easy enough to make it work without it.
> And it is also shorter ;)
> 
> —
> Reply to this email directly or view it on GitHub.
>